src/lib/comms/sol-bluetooth-impl-linux.c

```diff
@@ -0,0 +1,683 @@
+        device_properties, device_property_changed, ctx);
+    SOL_INT_CHECK_GOTO(r, < 0, error_map);
+
+    r = sol_ptr_vector_append(&ctx->devices, d);
+    SOL_INT_CHECK_GOTO(r, < 0, error_append);
+
+    return;
+
+error_append:
+    sol_bus_unmap_cached_properties(ctx->bluez, device_properties, ctx);
+error_map:
+    free(d->path);
+error_dup:
+    free(d);
+}
+
+static void
+device_removed(void *data, const char *path)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+
+    d = find_device_by_path(ctx, path);
+    SOL_NULL_CHECK(d);
+
+    destroy_device(d);
+    sol_ptr_vector_remove(&ctx->devices, d);
+}
+
+static const struct sol_bus_interfaces interfaces[] = {
+    { .name = "org.bluez.Adapter1",
+      .appeared = adapter_appeared,
+      .removed = adapter_removed },
+    { .name = "org.bluez.Device1",
+      .appeared = device_appeared,
+      .removed = device_removed },
+    { NULL }
+};
+
+SOL_API struct sol_bt_conn
+*sol_bt_conn_ref(struct sol_bt_conn *conn)
+{
+    SOL_NULL_CHECK(conn, NULL);
+
+    conn->ref++;
+    return conn;
+}
+
+static void
+destroy_conn(struct sol_bt_conn *conn)
+{
+
+}
+
+SOL_API void sol_bt_conn_unref(struct sol_bt_conn *conn)
+{
+    if (!conn)
+        return;
+
+    conn->ref--;
+    if (conn->ref > 0)
+        return;
+
+    destroy_conn(conn);
+}
+
+static void
+bluez_service_connected(void *data, const char *unique)
+{
+    struct context *ctx = data;
+
+    SOL_DBG("BlueZ service connected, unique %s", unique);
+
+    sol_ptr_vector_init(&ctx->devices);
+
+    sol_bus_watch_interfaces(ctx->bluez, interfaces, ctx);
+}
+
+static void
+bluez_service_disconnected(void *data)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+    uint16_t idx;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&ctx->devices, d, idx) {
+        destroy_device(d);
+    }
+    sol_ptr_vector_clear(&ctx->devices);
+
+    sol_bus_remove_interfaces_watch(ctx->bluez, interfaces, ctx);
+
+    if (ctx->enabled_callback) {
+            ctx->enabled_callback(ctx->enabled_user_data, false);
+    }
+}
+
+static int
+watch_bluez(void)
+{
+    int r;
+
+    if (context.bluez)
+        goto watch;
+
+    context.system_bus = sol_bus_get(NULL);
+    SOL_NULL_CHECK(context.system_bus, -EINVAL);
+
+    context.bluez = sol_bus_client_new(context.system_bus, "org.bluez");
+    SOL_NULL_CHECK(context.bluez, -EINVAL);
+
+watch:
+    r = sol_bus_client_set_connect_handler(context.bluez,
+        bluez_service_connected, &context);
+    SOL_INT_CHECK(r, < 0, -EINVAL);
+
+    r = sol_bus_client_set_disconnect_handler(context.bluez,
+        bluez_service_disconnected,
+        &context);
+    SOL_INT_CHECK(r, < 0, -EINVAL);
+
+    return 0;
+}
+
+SOL_API int
+sol_bt_set_connection_callbacks(void *(connected)(void *user_data,
+        struct sol_bt_conn *conn),
+    void *(disconnected)(void *user_data,
+        struct sol_bt_conn *conn),
+    const void *user_data)
```

Comment:
so why provide this? from the .h I was clueless why this was required and what was the use, expected to at least get a hint  here :-)